Check for nil in cPlayer:OpenWindow and close windows on reload

Seadragon91 opened the pull request:
Fixes #3727 

Also fixed a crash, if the player has a open window that has been created by a plugin. Fixed by closing window from all players on command reload.

madmaxoft commented:
I think a better solution would be:
 - Check in `OpenWindow` C++ code, if the passed object is a `nullptr` (what to do then? close current?)
 - Modify the `OpenWindow` C++ signature to use a reference rather than a pointer, tolua generates proper bindings then, including the `nil` check.

As for closing all the windows, that should not be necessary, the window object should be freed by Lua's GC and that should close the windows client-side. Definitely this is the wrong kind of fix - fixing the result, rather than the cause.

Seadragon91 commented:
For the lua window I read in [LuaWindow](https://github.com/cuberite/cuberite/blob/master/src/Bindings/LuaWindow.cpp) that it will keep a ref to itself to avoid being removed by GC. This ref will be available as long there is a player that has the window open.

madmaxoft commented:
It will keep a **Lua** reference. That means that Lua will still GC it when the `cLuaState` is closed. Perhaps the only problem is that the `cPlayer` objects aren't notified when that happens (I don't know, needs testing). So the closing would be better suited into the `cLuaWindow` destructor.

Seadragon91 commented:
Reworked it:
- Added overloaded method of OpenWindow accepting a ref to the window
- Closing cLuaWindow in destructor

Boths tested and works fine. Also tested if the cLuaWindow is still valid, if a ref is kept to open the same later again.

Seadragon91 commented:
Added missing `false` (a_CanRefuse) to `CloseWindow`

Seadragon91 commented:
Okay for merge?

madmaxoft commented on src/Entities/Player.h:
```diff
@@ -226,11 +226,14 @@ class cPlayer :
+	// tolua_begin
+
+	/** Calls OpenWindow with the pointer of it. This is exported by tolua and it adds the nil check that we want */
+	void OpenWindow(cWindow & a_Window) { OpenWindow(&a_Window); }
```
Why not remove the pointer-based version and just keep this one?

Seadragon91 replied:
Because of additional 10 line changes, change pointer to ref in all calls, I can change it, if you want :)

Seadragon91 replied:
Fixed all calls now, just search and replace :)

madmaxoft commented on src/Entities/Player.h:
```diff
@@ -228,8 +228,8 @@ class cPlayer :
 
-	/** Opens the specified window; closes the current one first using CloseWindow() */
-	void OpenWindow(cWindow * a_Window);
+	/** Opens the specified window; closes the current one first using CloseWindow(). This is exported by tolua and it adds the nil check we want */
```
I don't think the extra comment is necessary. It's something you've just fixed and no-one will be interested in it anymore. Export to Lua is visible because of the `tolua` comment just above.

Seadragon91 replied:
Removed it.
